src/providers/aws/mock_tests.rs

```diff
@@ -86,3 +86,88 @@ fn test_aws_attributes() {
     mockito::reset();
     provider.attributes().unwrap_err();
 }
+
+#[test]
+fn test_aws_imds_versions() {
+    let instance_id = "test-instance-id";
+    let instance_type = "test-instance-type";
+    let ipv4_local = "test-ipv4-local";
+    let ipv4_public = "test-ipv4-public";
+    let availability_zone = "test-availability-zone";
+    let hostname = "test-hostname";
+    let public_hostname = "test-public-hostname";
+    let instance_id_doc = r#"{"region": "test-region"}"#;
+    let region = "test-region";
+
+    let attributes = maplit::hashmap! {
+        format!("{}_INSTANCE_ID", aws::ENV_PREFIX) => instance_id.to_string(),
+        format!("{}_INSTANCE_TYPE", aws::ENV_PREFIX) => instance_type.to_string(),
+        format!("{}_IPV4_LOCAL", aws::ENV_PREFIX) => ipv4_local.to_string(),
+        format!("{}_IPV4_PUBLIC", aws::ENV_PREFIX) => ipv4_public.to_string(),
+        format!("{}_AVAILABILITY_ZONE", aws::ENV_PREFIX) => availability_zone.to_string(),
+        format!("{}_HOSTNAME", aws::ENV_PREFIX) => hostname.to_string(),
+        format!("{}_PUBLIC_HOSTNAME", aws::ENV_PREFIX) => public_hostname.to_string(),
+        format!("{}_REGION", aws::ENV_PREFIX) => region.to_string(),
+    };
+
+    let endpoints = maplit::btreemap! {
+        "/meta-data/instance-id" => instance_id,
+        "/meta-data/instance-type" => instance_type,
+        "/meta-data/local-ipv4" => ipv4_local,
+        "/meta-data/public-ipv4" => ipv4_public,
+        "/meta-data/placement/availability-zone" => availability_zone,
+        "/meta-data/hostname" => hostname,
+        "/meta-data/public-hostname" => public_hostname,
+        "/dynamic/instance-identity/document" => instance_id_doc,
+    };
+
+    // first test imdsv1
```

Comment:
I think there is something fishy with the test as it takes a long time to complete, but I'm not sure why.

It would be good to start scoping the v1 and v2 sublogic into their own `{ ... }` scopes to make sure there isn't anything leaking from the former to the latter, that is:
```rust
 // first test imdsv1
{
let mut mocks = ...
...
}

// then test imdsv2
{
let token = ...
...
}
```

Reply:
I feel like the long time comes from retries on the tested failing PUT requests. By default there will be 10 retries and 5 sec maximum back off time. But still the long running time is :(

Tried to time the test:
 - imdsv1 only: 91 sec
 - imdsv2 only: 42 sec
 - imdsv1 and imdsv2: 132 sec

But from AWS the `afterburn --cmdline --attributes output` finished almost instantaneously.

Reply:
Ah, it could be that then. If you have a look at other platform mocks (e.g. `gcp`), you'll see we tweak down the number of retries. We should probably do the same for `aws` (possibly in another PR).

Reply:
Tested with `.max_retries(0)`, the test succeeded immediately.. nice :+1: 
Updating that bit as well

Reply:
```diff
 impl AwsProvider {
     pub fn try_new() -> Result<AwsProvider> {
+        #[cfg(test)]
+        let mut client = retry::Client::try_new()?.max_retries(0).return_on_404(true);
+        #[cfg(not(test))]
         let mut client = retry::Client::try_new()?.return_on_404(true);
         let token = AwsProvider::fetch_imdsv2_token(client.clone());

```

Reply:
Ah, nice. Can we keep the `max_retries` in here instead of making that conditional in `try_new()`?

Reply:
Yep, updating :+1: